File: ruby/sum-of-multiples/sum_of_multiples.rb
# frozen_string_literal: false

require 'set'

# https://exercism.org/tracks/ruby/exercises/sum-of-multiples
# Sum of Multiples exercise
class SumOfMultiples
  def initialize(*args)
    @magical_items = args

    @multiples = Set[]
  end

  def to(base_value)
    return 0 if base_value.zero?

    @magical_items.each do |item|
      get_multiples item, base_value
    end

    @multiples.sum
  end

  def get_multiples(factor, limit)
    (1...limit).each do |multiple|
      @multiples.add(multiple) if multiple.modulo(factor).zero?
    end
  end

  private :get_multiples
end

File: lua/sum-of-multiples/sum-of-multiples.lua
return function(numbers)
    local function get_multiples(factor, limit)
        local multiples = {}

        for multiple = 1, limit - 1 do
            if multiple % factor == 0 then
                table.insert(multiples, multiple)
            end
        end

        return multiples
    end

    local function to(base_value)
        local sum = 0

        if base_value == 0 then
            return sum
        end

        local hash = {}
        local unique = {}
        local multiples = {}

        for _, item in pairs(numbers) do
            for _, number in ipairs(get_multiples(item, base_value)) do
                table.insert(multiples, number)
            end
        end

        for _, value in ipairs(multiples) do
            hash[value] = true
        end

        for key, _ in pairs(hash) do
            table.insert(unique, key)
        end

        for _, value in ipairs(unique) do
            sum = sum + value
        end

        return sum
    end

    return {
        to = to,
    }
end
